internal/v4/archive.go

```diff
@@ -327,10 +344,18 @@ func (h *ReqHandler) addEntity(id *router.ResolvedURL, r io.ReadSeeker, blobName
 	}
 	ch, err := charm.ReadCharmArchiveFromReader(readerAt, contentLength)
 	if err != nil {
-		return errgo.Notef(err, "cannot read charm archive")
+		switch errgo.Cause(err) {
```

Comment:
Maybe use a sperate function for handling this error, since the same logic applies to charms and bundles?